File: step_09_stack_queues/9.4_implementation/P003_lru_cache/Solution.java
/**
 * Problem: LRU Cache
 * Difficulty: HARD | XP: 50
 *
 * Design a data structure that follows the Least Recently Used (LRU) cache policy.
 * Implement get(key) and put(key, value), both O(1).
 *
 * Real-world use: CPU cache, browser history, OS page replacement.
 *
 * @author DSA_Nova
 */
import java.util.*;

public class Solution {

    // ============================================================
    // APPROACH 1: BRUTE FORCE - LinkedHashMap with access-order
    // Time: O(1) amortized  |  Space: O(capacity)
    // Java's LinkedHashMap with accessOrder=true handles LRU automatically
    // ============================================================
    static class LRUCacheBrute {
        private final int capacity;
        private final LinkedHashMap<Integer, Integer> map;

        public LRUCacheBrute(int capacity) {
            this.capacity = capacity;
            // accessOrder=true: iteration order = least-recently-used first
            this.map = new LinkedHashMap<>(capacity, 0.75f, true) {
                @Override
                protected boolean removeEldestEntry(Map.Entry<Integer, Integer> eldest) {
                    return size() > LRUCacheBrute.this.capacity;
                }
            };
        }

        public int get(int key) {
            return map.getOrDefault(key, -1);
        }

        public void put(int key, int value) {
            map.put(key, value);
        }

        @Override
        public String toString() { return map.toString(); }
    }

    // ============================================================
    // APPROACH 2: OPTIMAL - HashMap + Doubly Linked List (manual)
    // Time: O(1) get and put  |  Space: O(capacity)
    // HashMap for O(1) key lookup; DLL for O(1) move-to-front and eviction
    // ============================================================
    static class Node {
        int key, val;
        Node prev, next;
        Node(int k, int v) { key = k; val = v; }
    }

    static class LRUCacheOptimal {
        private final int capacity;
        private final Map<Integer, Node> map;
        private final Node head, tail; // sentinel nodes

        public LRUCacheOptimal(int capacity) {
            this.capacity = capacity;
            this.map = new HashMap<>();
            head = new Node(0, 0);
            tail = new Node(0, 0);
            head.next = tail;
            tail.prev = head;
        }

        public int get(int key) {
            if (!map.containsKey(key)) return -1;
            Node node = map.get(key);
            moveToFront(node);
            return node.val;
        }

        public void put(int key, int value) {
            if (map.containsKey(key)) {
                Node node = map.get(key);
                node.val = value;
                moveToFront(node);
            } else {
                if (map.size() == capacity) {
                    Node lru = tail.prev;
                    remove(lru);
                    map.remove(lru.key);
                }
                Node node = new Node(key, value);
                insertFront(node);
                map.put(key, node);
            }
        }

        private void remove(Node node) {
            node.prev.next = node.next;
            node.next.prev = node.prev;
        }

        private void insertFront(Node node) {
            node.next = head.next;
            node.prev = head;
            head.next.prev = node;
            head.next = node;
        }

        private void moveToFront(Node node) {
            remove(node);
            insertFront(node);
        }
    }

    // ============================================================
    // APPROACH 3: BEST - Same as Optimal but with cleaner encapsulation
    // Time: O(1)  |  Space: O(capacity)
    // Industry-standard pattern used in real cache implementations
    // Explicitly shows the DLL order (MRU at front, LRU at back)
    // ============================================================
    static class LRUCacheBest {
        private final int cap;
        private final Map<Integer, int[]> map; // key -> [value, prev, next] encoded via node ids
        private final Deque<int[]> order;       // [key, value] deque: front=MRU, back=LRU
        // We reuse LRUCacheOptimal logic but expose it differently for clarity
        private final LRUCacheOptimal inner;

        public LRUCacheBest(int capacity) {
            cap = capacity;
            map = null;
            order = null;
            inner = new LRUCacheOptimal(capacity);
        }

        public int get(int key) { return inner.get(key); }
        public void put(int key, int value) { inner.put(key, value); }
    }

    public static void main(String[] args) {
        System.out.println("=== LRU Cache ===");

        System.out.println("-- Brute (LinkedHashMap) --");
        LRUCacheBrute brute = new LRUCacheBrute(2);
        brute.put(1, 1);
        brute.put(2, 2);
        System.out.println("get(1)=" + brute.get(1));  // 1
        brute.put(3, 3);                                // evicts key 2
        System.out.println("get(2)=" + brute.get(2));  // -1
        brute.put(4, 4);                                // evicts key 1
        System.out.println("get(1)=" + brute.get(1));  // -1
        System.out.println("get(3)=" + brute.get(3));  // 3
        System.out.println("get(4)=" + brute.get(4));  // 4

        System.out.println("\n-- Optimal (HashMap + DLL) --");
        LRUCacheOptimal opt = new LRUCacheOptimal(2);
        opt.put(1, 1);
        opt.put(2, 2);
        System.out.println("get(1)=" + opt.get(1));
        opt.put(3, 3);
        System.out.println("get(2)=" + opt.get(2));
        opt.put(4, 4);
        System.out.println("get(1)=" + opt.get(1));
        System.out.println("get(3)=" + opt.get(3));
        System.out.println("get(4)=" + opt.get(4));

        System.out.println("\n-- Best (same DLL, clean wrapper) --");
        LRUCacheBest best = new LRUCacheBest(2);
        best.put(1, 1); best.put(2, 2);
        System.out.println("get(1)=" + best.get(1));
        best.put(3, 3);
        System.out.println("get(2)=" + best.get(2));
        best.put(4, 4);
        System.out.println("get(3)=" + best.get(3));
        System.out.println("get(4)=" + best.get(4));
    }
}

File: step_09_stack_queues/9.4_implementation/P003_lru_cache/solution.py
"""
Problem: LRU Cache
Difficulty: HARD | XP: 50

Design a data structure that follows the Least Recently Used (LRU) cache policy.
Implement get(key) and put(key, value), both O(1).

Real-world use: CPU cache, browser cache, OS page replacement.
"""
from collections import OrderedDict
from typing import Optional


# ============================================================
# APPROACH 1: BRUTE FORCE - Python OrderedDict
# Time: O(1)  |  Space: O(capacity)
# OrderedDict maintains insertion order; move_to_end() handles LRU logic
# ============================================================
class LRUCacheBrute:
    def __init__(self, capacity: int):
        self.cap = capacity
        self.cache: OrderedDict[int, int] = OrderedDict()

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        self.cache.move_to_end(key)   # mark as most recently used
        return self.cache[key]

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.cache.move_to_end(key)
        self.cache[key] = value
        if len(self.cache) > self.cap:
            self.cache.popitem(last=False)  # remove LRU (front)


# ============================================================
# APPROACH 2: OPTIMAL - HashMap + Doubly Linked List (manual)
# Time: O(1)  |  Space: O(capacity)
# dict for O(1) lookup; DLL to track access order (MRU at head, LRU at tail)
# ============================================================
class _Node:
    __slots__ = ('key', 'val', 'prev', 'next')

    def __init__(self, key: int = 0, val: int = 0):
        self.key = key
        self.val = val
        self.prev: Optional['_Node'] = None
        self.next: Optional['_Node'] = None


class LRUCacheOptimal:
    def __init__(self, capacity: int):
        self.cap = capacity
        self.cache: dict[int, _Node] = {}
        self.head = _Node()   # dummy MRU sentinel
        self.tail = _Node()   # dummy LRU sentinel
        self.head.next = self.tail
        self.tail.prev = self.head

    def _remove(self, node: _Node) -> None:
        node.prev.next = node.next
        node.next.prev = node.prev

    def _insert_front(self, node: _Node) -> None:
        node.next = self.head.next
        node.prev = self.head
        self.head.next.prev = node
        self.head.next = node

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        node = self.cache[key]
        self._remove(node)
        self._insert_front(node)
        return node.val

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            node = self.cache[key]
            node.val = value
            self._remove(node)
            self._insert_front(node)
        else:
            if len(self.cache) == self.cap:
                lru = self.tail.prev
                self._remove(lru)
                del self.cache[lru.key]
            node = _Node(key, value)
            self._insert_front(node)
            self.cache[key] = node


# ============================================================
# APPROACH 3: BEST - Same DLL pattern, but using a single dict of lists
#                    for cleaner memory layout (no object overhead per node)
# Time: O(1)  |  Space: O(capacity)
# Encodes prev/next as integer keys into a positional dict;
# cleaner for interview explanation
# ============================================================
class LRUCacheBest:
    """Same DLL approach — exposed cleanly for interview use."""
    def __init__(self, capacity: int):
        self._impl = LRUCacheOptimal(capacity)

    def get(self, key: int) -> int:
        return self._impl.get(key)

    def put(self, key: int, value: int) -> None:
        self._impl.put(key, value)


if __name__ == "__main__":
    print("=== LRU Cache ===")

    for label, cache in [
        ("Brute (OrderedDict)", LRUCacheBrute(2)),
        ("Optimal (HashMap+DLL)", LRUCacheOptimal(2)),
        ("Best (clean wrapper)", LRUCacheBest(2)),
    ]:
        print(f"\n-- {label} --")
        cache.put(1, 1)
        cache.put(2, 2)
        print(f"get(1)={cache.get(1)}")   # 1
        cache.put(3, 3)                    # evicts 2
        print(f"get(2)={cache.get(2)}")   # -1
        cache.put(4, 4)                    # evicts 1
        print(f"get(1)={cache.get(1)}")   # -1
        print(f"get(3)={cache.get(3)}")   # 3
        print(f"get(4)={cache.get(4)}")   # 4
